File: backend/internal/service/collection_service.go
package service

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/kenyamamoto/pokelingual/backend/internal/apperror"
	"github.com/kenyamamoto/pokelingual/backend/internal/domain"
	"github.com/kenyamamoto/pokelingual/backend/internal/model"
)

// CollectionService provides access to the user's captured Pokemon collection.
type CollectionService struct {
	repo           domain.UserPokemonRepository
	pokemonFetcher domain.PokemonFetcher
}

// CollectionEntry represents a single Pokemon in the user's collection list.
type CollectionEntry struct {
	PokemonID     int     `json:"pokemon_id"`
	NameEN        string  `json:"name_en"`
	NameJA        string  `json:"name_ja"`
	SpriteURL     string  `json:"sprite_url"`
	Status        string  `json:"status"`
	TotalCaptures int     `json:"total_captures"`
	BestScore     float64 `json:"best_score"`
}

// PokemonDetailResponse is the API response for a captured Pokemon's details.
type PokemonDetailResponse struct {
	model.UserPokemon
	NameEN        string                 `json:"name_en"`
	NameJA        string                 `json:"name_ja"`
	DescriptionEN string                 `json:"description_en"`
	DescriptionJA string                 `json:"description_ja"`
	SpriteURL     string                 `json:"sprite_url"`
	FlavorTexts   []model.FlavorTextPair `json:"flavor_texts,omitempty"`
}

// NewCollectionService creates a new CollectionService.
func NewCollectionService(repo domain.UserPokemonRepository, pokemonFetcher domain.PokemonFetcher) *CollectionService {
	return &CollectionService{
		repo:           repo,
		pokemonFetcher: pokemonFetcher,
	}
}

// GetCollection returns all discovered Pokemon for the given user.
func (s *CollectionService) GetCollection(ctx context.Context, uid string) ([]CollectionEntry, error) {
	pokemons, err := s.repo.GetCollection(ctx, uid)
	if err != nil {
		return nil, fmt.Errorf("getting collection: %w", err)
	}

	entries := make([]CollectionEntry, 0, len(pokemons))
	for _, up := range pokemons {
		pokemon, err := s.pokemonFetcher.GetPokemonByID(ctx, up.PokemonID)
		if err != nil {
			slog.Warn("failed to fetch pokemon data, skipping", "pokemon_id", up.PokemonID, "error", err)
			continue
		}
		entries = append(entries, CollectionEntry{
			PokemonID:     up.PokemonID,
			NameEN:        pokemon.NameEN,
			NameJA:        pokemon.NameJA,
			SpriteURL:     pokemon.SpriteURL,
			Status:        up.Status,
			TotalCaptures: up.TotalCaptures,
			BestScore:     up.BestScore,
		})
	}

	return entries, nil
}

// GetPokemonDetail returns the full details of a specific captured Pokemon.
func (s *CollectionService) GetPokemonDetail(ctx context.Context, uid string, pokemonID int) (*PokemonDetailResponse, error) {
	userPokemon, err := s.repo.GetPokemon(ctx, uid, pokemonID)
	if err != nil {
		return nil, fmt.Errorf("getting user pokemon: %w", err)
	}

	pokemon, err := s.pokemonFetcher.GetPokemonByID(ctx, pokemonID)
	if err != nil {
		return nil, apperror.NewExternalServiceError("PokeAPI", err)
	}

	return &PokemonDetailResponse{
		UserPokemon:   *userPokemon,
		NameEN:        pokemon.NameEN,
		NameJA:        pokemon.NameJA,
		DescriptionEN: pokemon.DescriptionEN,
		DescriptionJA: pokemon.DescriptionJA,
		SpriteURL:     pokemon.SpriteURL,
		FlavorTexts:   pokemon.FlavorTexts,
	}, nil
}

File: backend/src/service/collection-service.ts
import { ExternalServiceError } from "../apperror/apperror.js";
import type { PokemonFetcher, UserPokemonRepository } from "../domain/interfaces.js";
import type { FlavorTextPair, UserPokemon } from "../types/index.js";

/** 図鑑一覧の 1 エントリ。ユーザ実績と表示用のポケモン基本情報を含む。 */
export interface CollectionEntry {
  pokemon_id: number;
  name_en: string;
  name_ja: string;
  sprite_url: string;
  status: string;
  total_captures: number;
  best_score: number;
}

/** ポケモン詳細レスポンス。ユーザ実績と PokeAPI 情報を合成して返す。 */
export interface PokemonDetailResponse extends UserPokemon {
  name_en: string;
  name_ja: string;
  description_en: string;
  description_ja: string;
  sprite_url: string;
  types: string[];
  height: number;
  weight: number;
  flavor_texts?: FlavorTextPair[];
}

/** 図鑑コレクションの取得・整形ロジックを束ねるサービス。 */
export class CollectionService {
  private repo: UserPokemonRepository;
  private pokemonFetcher: PokemonFetcher;

  constructor(repo: UserPokemonRepository, pokemonFetcher: PokemonFetcher) {
    this.repo = repo;
    this.pokemonFetcher = pokemonFetcher;
  }

  /** ユーザの図鑑一覧を取得し、PokeAPI からのメタ情報を付与して返す。 */
  async getCollection(uid: string): Promise<CollectionEntry[]> {
    const pokemons = await this.repo.getCollection(uid);
    const entries: CollectionEntry[] = [];

    for (const up of pokemons) {
      try {
        const pokemon = await this.pokemonFetcher.getPokemonByID(up.pokemon_id);
        entries.push({
          pokemon_id: up.pokemon_id,
          name_en: pokemon.name_en,
          name_ja: pokemon.name_ja,
          sprite_url: pokemon.sprite_url,
          status: up.status,
          total_captures: up.total_captures,
          best_score: up.best_score,
        });
      } catch (err) {
        console.warn(`failed to fetch pokemon data, skipping pokemon_id=${up.pokemon_id}`, err);
      }
    }

    return entries;
  }

  /** 特定ポケモンのユーザ実績と PokeAPI 詳細を合成して返す。 */
  async getPokemonDetail(uid: string, pokemonID: number): Promise<PokemonDetailResponse> {
    const userPokemon = await this.repo.getPokemon(uid, pokemonID);

    let pokemon;
    try {
      pokemon = await this.pokemonFetcher.getPokemonByID(pokemonID);
    } catch (err) {
      throw new ExternalServiceError("PokeAPI", err as Error);
    }

    return {
      ...userPokemon,
      name_en: pokemon.name_en,
      name_ja: pokemon.name_ja,
      description_en: pokemon.description_en,
      description_ja: pokemon.description_ja,
      sprite_url: pokemon.sprite_url,
      types: pokemon.types,
      height: pokemon.height,
      weight: pokemon.weight,
      flavor_texts: pokemon.flavor_texts,
    };
  }
}
